src/client.ts

```diff
@@ -1,18 +1,188 @@
 import fetch from 'isomorphic-unfetch';
-import { isFunction, isObject, isString } from '../utils/helper';
 import {
   ClientObjectProps,
-  ReturnType,
-  RequestConfig,
+  ResponseType,
   Callback,
-  MethodReturn,
   Part,
   Upstash,
   Bit,
   ZSetNumber,
   EdgeCacheType,
+  Pipeline,
 } from './type';
 
+type CustomArgsFn = (...args: any[]) => any[];
+
+type OperationValue =
+  | number
+  | readonly [number, ...(number | boolean | 'spread' | CustomArgsFn)[]]
+  | ((options: ClientObjectProps, ...args: any[]) => any);
+
+type RequestConfig = {
+  options: ClientObjectProps;
+  edge?: boolean;
+};
+
+function parseOptions(
+  url?: string | ClientObjectProps,
+  token?: string,
+  edgeUrl?: string | null,
+  readFromEdge?: boolean
+): ClientObjectProps {
+  if (typeof url === 'object')
+    return parseOptions(url.url, url.token, url.edgeUrl, url.readFromEdge);
+  // Try to auto fill from env variables
+  if (!url) url = process.env.UPSTASH_REDIS_REST_URL;
+  if (!token) token = process.env.UPSTASH_REDIS_REST_TOKEN;
+  if (edgeUrl === undefined) edgeUrl = process.env.UPSTASH_REDIS_EDGE_URL;
+  readFromEdge = readFromEdge ?? !!edgeUrl;
+
+  if (readFromEdge && !edgeUrl) {
+    throw new Error(`readFromEdge is set to true but the Edge Url is missing`);
+  }
+  if (!url) throw new Error('REST Url is missing');
```

Comment:
I had to comment out these two `if` checks during my test, because any import from `@upstash/redis` such as 
```js
import {auth, ping, get} from '@upstash/redis';
```
was failing with `REST Url is missing` error. 

It seems the reason is `Upstash()` call in `index.ts`.

Reply:
Additionally, `src/` in the middle of `dist/main/src/index.js` and `dist/module/src/index.js` should be removed in `package.json`; 

```json
  "main": "dist/main/index.js",
  "types": "dist/main/index.d.ts",
  "module": "dist/module/index.js",
```

`src/` sub-dir is not created anymore in `dist` folder.